File: Practica 1/r/01_limpieza_datos.R
suppressPackageStartupMessages({
  library(readr)
  library(dplyr)
  library(stringr)
  library(lubridate)
})

args <- commandArgs(trailingOnly = TRUE)

ruta_csv <- ifelse(
  length(args) >= 1,
  args[1],
  "../ventas_online_2025.csv"
)

salida_csv <- ifelse(
  length(args) >= 2,
  args[2],
  "ventas_online_2025_limpio.csv"
)

reporte_txt <- ifelse(
  length(args) >= 3,
  args[3],
  "reporte_limpieza.txt"
)

limpiar_numero <- function(x) {
  x <- as.character(x)
  x <- str_trim(x)
  x <- str_replace_all(x, fixed("$"), "")
  x <- str_replace_all(x, fixed('"'), "")
  x <- str_replace_all(x, fixed("'"), "")

  resultado <- sapply(x, function(v) {
    if (is.na(v) || v == "") {
      return(NA_real_)
    }

    if (str_detect(v, ",") && !str_detect(v, fixed("."))) {
      v <- str_replace_all(v, ",", ".")
    } else if (str_detect(v, ",") && str_detect(v, fixed("."))) {
      v <- str_replace_all(v, ",", "")
    }

    suppressWarnings(as.numeric(v))
  })

  as.numeric(resultado)
}

normalizar_genero <- function(x) {
  x <- as.character(x)
  x <- str_to_lower(str_trim(x))
  x <- str_replace_all(x, "í", "i")

  case_when(
    x %in% c("0", "0.0", "masculino") ~ 0,
    x %in% c("1", "1.0", "femenino") ~ 1,
    TRUE ~ NA_real_
  )
}

normalizar_si_no <- function(x) {
  x <- as.character(x)
  x <- str_to_lower(str_trim(x))
  x <- str_replace_all(x, "í", "i")

  case_when(
    x %in% c("0", "0.0", "no") ~ 0,
    x %in% c("1", "1.0", "si") ~ 1,
    TRUE ~ NA_real_
  )
}

moda <- function(x) {
  x <- x[!is.na(x)]
  if (length(x) == 0) {
    return(NA_real_)
  }

  valores <- unique(x)
  valores[which.max(tabulate(match(x, valores)))]
}

df <- read_csv(ruta_csv, show_col_types = FALSE)

filas_iniciales <- nrow(df)

names(df) <- str_trim(names(df))
names(df)[names(df) == "Boletín"] <- "Boletin"

requeridas <- c(
  "Id_cliente", "Edad", "Genero", "Venta_total",
  "N_Compras", "FechaCompra", "MontoCompra",
  "MetodoPago", "Tiempo", "Navegador",
  "Boletin", "Vale"
)

faltan <- setdiff(requeridas, names(df))
if (length(faltan) > 0) {
  stop(
    paste(
      "Faltan columnas requeridas:",
      paste(faltan, collapse = ", ")
    )
  )
}

duplicados <- sum(duplicated(df))
df <- distinct(df)

df$Genero <- normalizar_genero(df$Genero)
df$Boletin <- normalizar_si_no(df$Boletin)
df$Vale <- normalizar_si_no(df$Vale)

df$Venta_total <- limpiar_numero(df$Venta_total)
df$MontoCompra <- limpiar_numero(df$MontoCompra)

for (columna in c(
  "Id_cliente", "Edad", "N_Compras",
  "MetodoPago", "Tiempo", "Navegador"
)) {
  df[[columna]] <- suppressWarnings(as.numeric(df[[columna]]))
}

df$FechaCompra <- suppressWarnings(
  parse_date_time(
    as.character(df$FechaCompra),
    orders = c(
      "Y-m-d",
      "Y/m/d",
      "d-m-Y",
      "d/m/Y",
      "m/d/Y"
    )
  )
)

df$FechaCompra <- as.Date(df$FechaCompra)

# Reglas de validez
df$Edad[df$Edad < 0 | df$Edad > 100] <- NA
df$Venta_total[df$Venta_total < 0] <- NA
df$MontoCompra[df$MontoCompra < 0] <- NA

df$MetodoPago[!df$MetodoPago %in% c(0, 1, 2)] <- NA
df$Navegador[!df$Navegador %in% c(0, 1, 2, 3, 4)] <- NA

faltantes_antes <- sapply(df, function(x) sum(is.na(x)))

# Imputación categórica con moda
for (columna in c(
  "Genero", "MetodoPago", "Navegador",
  "Boletin", "Vale"
)) {
  valor_moda <- moda(df[[columna]])

  if (!is.na(valor_moda)) {
    df[[columna]][is.na(df[[columna]])] <- valor_moda
  }
}

# Imputación numérica con mediana
for (columna in c(
  "Edad", "Venta_total", "N_Compras",
  "MontoCompra", "Tiempo"
)) {
  mediana <- median(df[[columna]], na.rm = TRUE)
  df[[columna]][is.na(df[[columna]])] <- mediana
}

fechas_invalidas <- sum(is.na(df$FechaCompra))
df <- df |>
  filter(!is.na(FechaCompra))

# Enteros finales
for (columna in c(
  "Id_cliente", "Genero", "N_Compras",
  "MetodoPago", "Navegador", "Boletin", "Vale"
)) {
  df[[columna]] <- as.integer(round(df[[columna]]))
}

faltantes_despues <- sapply(df, function(x) sum(is.na(x)))

write_csv(df, salida_csv)

reporte <- c(
  "REPORTE DE LIMPIEZA",
  "===================",
  paste("Filas originales:", filas_iniciales),
  paste("Duplicados eliminados:", duplicados),
  paste("Fechas inválidas eliminadas:", fechas_invalidas),
  paste("Filas finales:", nrow(df)),
  "",
  "VALORES FALTANTES / INVÁLIDOS DETECTADOS ANTES DE IMPUTAR"
)

for (nombre in names(faltantes_antes)) {
  reporte <- c(
    reporte,
    paste("-", nombre, ":", faltantes_antes[[nombre]])
  )
}

reporte <- c(
  reporte,
  "",
  "VALORES FALTANTES DESPUÉS DE LA LIMPIEZA"
)

for (nombre in names(faltantes_despues)) {
  reporte <- c(
    reporte,
    paste("-", nombre, ":", faltantes_despues[[nombre]])
  )
}

reporte <- c(
  reporte,
  "",
  "CRITERIOS UTILIZADOS",
  "- Edad válida: 0 a 100 años.",
  "- Genero válido: 0 = Masculino, 1 = Femenino.",
  "- MetodoPago válido: 0, 1 o 2.",
  "- Navegador válido: 0, 1, 2, 3 o 4.",
  "- Boletin y Vale válidos: 0 o 1.",
  "- Montos negativos se consideran inválidos.",
  "- Numéricos faltantes/inválidos: mediana.",
  "- Categóricos faltantes/inválidos: moda.",
  "- Duplicados exactos: eliminados."
)

writeLines(reporte, reporte_txt, useBytes = TRUE)

message("Archivo limpio: ", normalizePath(salida_csv))
message("Reporte: ", normalizePath(reporte_txt))

File: Practica 1/python/01_limpieza_datos.py
from __future__ import annotations

import argparse
from pathlib import Path
import numpy as np
import pandas as pd


def limpiar_numero(valor):
    if pd.isna(valor):
        return np.nan

    texto = str(valor).strip()
    texto = texto.replace("$", "").replace('"', "").replace("'", "").strip()

    if texto == "":
        return np.nan

    # Si únicamente hay coma, se considera separador decimal.
    if "," in texto and "." not in texto:
        texto = texto.replace(",", ".")
    elif "," in texto and "." in texto:
        texto = texto.replace(",", "")

    try:
        return float(texto)
    except ValueError:
        return np.nan


def normalizar_genero(valor):
    if pd.isna(valor):
        return np.nan

    texto = str(valor).strip().lower().replace("í", "i")

    if texto in {"0", "0.0", "masculino"}:
        return 0
    if texto in {"1", "1.0", "femenino"}:
        return 1

    return np.nan


def normalizar_si_no(valor):
    if pd.isna(valor):
        return np.nan

    texto = str(valor).strip().lower().replace("í", "i")

    if texto in {"0", "0.0", "no"}:
        return 0
    if texto in {"1", "1.0", "si"}:
        return 1

    return np.nan


def convertir_fecha(valor):
    if pd.isna(valor):
        return pd.NaT

    texto = str(valor).strip()

    formatos = [
        "%Y-%m-%d",
        "%Y/%m/%d",
        "%d-%m-%Y",
        "%d/%m/%Y",
        "%m/%d/%Y",
    ]

    for formato in formatos:
        try:
            return pd.to_datetime(texto, format=formato)
        except (ValueError, TypeError):
            pass

    return pd.NaT


def moda(serie):
    resultado = serie.mode(dropna=True)
    return resultado.iloc[0] if not resultado.empty else np.nan


def limpiar_datos(ruta_csv: str, salida_csv: str, reporte_txt: str):
    df = pd.read_csv(ruta_csv)

    filas_iniciales = len(df)

    # 1. Normalización de nombres de columnas
    df.columns = [c.strip() for c in df.columns]
    df = df.rename(columns={"Boletín": "Boletin"})

    columnas_requeridas = [
        "Id_cliente", "Edad", "Genero", "Venta_total", "N_Compras",
        "FechaCompra", "MontoCompra", "MetodoPago", "Tiempo",
        "Navegador", "Boletin", "Vale"
    ]

    faltantes = [c for c in columnas_requeridas if c not in df.columns]
    if faltantes:
        raise ValueError(f"Faltan columnas requeridas: {faltantes}")

    # 2. Duplicados
    duplicados = int(df.duplicated().sum())
    df = df.drop_duplicates().copy()

    # 3. Conversión / estandarización
    df["Genero"] = df["Genero"].apply(normalizar_genero)
    df["Boletin"] = df["Boletin"].apply(normalizar_si_no)
    df["Vale"] = df["Vale"].apply(normalizar_si_no)

    for columna in ["Venta_total", "MontoCompra"]:
        df[columna] = df[columna].apply(limpiar_numero)

    for columna in ["Id_cliente", "Edad", "N_Compras", "MetodoPago", "Tiempo", "Navegador"]:
        df[columna] = pd.to_numeric(df[columna], errors="coerce")

    df["FechaCompra"] = df["FechaCompra"].apply(convertir_fecha)

    # 4. Validación de reglas
    df.loc[(df["Edad"] < 0) | (df["Edad"] > 100), "Edad"] = np.nan
    df.loc[df["Venta_total"] < 0, "Venta_total"] = np.nan
    df.loc[df["MontoCompra"] < 0, "MontoCompra"] = np.nan

    df.loc[~df["MetodoPago"].isin([0, 1, 2]), "MetodoPago"] = np.nan
    df.loc[~df["Navegador"].isin([0, 1, 2, 3, 4]), "Navegador"] = np.nan

    faltantes_antes = df.isna().sum().to_dict()

    # 5. Imputación de faltantes / inválidos
    categoricas = ["Genero", "MetodoPago", "Navegador", "Boletin", "Vale"]
    for columna in categoricas:
        valor_moda = moda(df[columna])
        if not pd.isna(valor_moda):
            df[columna] = df[columna].fillna(valor_moda)

    numericas = ["Edad", "Venta_total", "N_Compras", "MontoCompra", "Tiempo"]
    for columna in numericas:
        mediana = df[columna].median()
        df[columna] = df[columna].fillna(mediana)

    # Si alguna fecha no pudo convertirse, se elimina la fila porque
    # no puede utilizarse correctamente en el análisis mensual.
    fechas_invalidas = int(df["FechaCompra"].isna().sum())
    df = df.dropna(subset=["FechaCompra"]).copy()

    # 6. Tipos finales
    for columna in [
        "Id_cliente", "Genero", "N_Compras",
        "MetodoPago", "Navegador", "Boletin", "Vale"
    ]:
        df[columna] = df[columna].round().astype(int)

    df["FechaCompra"] = df["FechaCompra"].dt.strftime("%Y-%m-%d")

    # 7. Guardar
    salida = Path(salida_csv)
    salida.parent.mkdir(parents=True, exist_ok=True)
    df.to_csv(salida, index=False, encoding="utf-8-sig")

    faltantes_despues = df.isna().sum().to_dict()

    reporte = [
        "REPORTE DE LIMPIEZA",
        "===================",
        f"Filas originales: {filas_iniciales}",
        f"Duplicados eliminados: {duplicados}",
        f"Fechas inválidas eliminadas: {fechas_invalidas}",
        f"Filas finales: {len(df)}",
        "",
        "VALORES FALTANTES / INVÁLIDOS DETECTADOS ANTES DE IMPUTAR",
    ]

    for columna, cantidad in faltantes_antes.items():
        reporte.append(f"- {columna}: {cantidad}")

    reporte.extend([
        "",
        "VALORES FALTANTES DESPUÉS DE LA LIMPIEZA",
    ])

    for columna, cantidad in faltantes_despues.items():
        reporte.append(f"- {columna}: {cantidad}")

    reporte.extend([
        "",
        "CRITERIOS UTILIZADOS",
        "- Edad válida: 0 a 100 años.",
        "- Genero válido: 0 = Masculino, 1 = Femenino.",
        "- MetodoPago válido: 0, 1 o 2.",
        "- Navegador válido: 0, 1, 2, 3 o 4.",
        "- Boletin y Vale válidos: 0 o 1.",
        "- Montos negativos se consideran inválidos.",
        "- Numéricos faltantes/inválidos: mediana.",
        "- Categóricos faltantes/inválidos: moda.",
        "- Duplicados exactos: eliminados.",
    ])

    Path(reporte_txt).write_text("\n".join(reporte), encoding="utf-8")

    print(f"Archivo limpio: {salida.resolve()}")
    print(f"Reporte: {Path(reporte_txt).resolve()}")
    print(f"Filas finales: {len(df)}")


def main():
    parser = argparse.ArgumentParser(description="Limpieza de ventas online 2025")
    parser.add_argument("csv", nargs="?", default="../ventas_online_2025.csv")
    parser.add_argument(
        "--salida",
        default="ventas_online_2025_limpio.csv"
    )
    parser.add_argument(
        "--reporte",
        default="reporte_limpieza.txt"
    )
    args = parser.parse_args()

    limpiar_datos(args.csv, args.salida, args.reporte)


if __name__ == "__main__":
    main()
